src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+            PcntCountMode::Increment => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_INCREASE,
+            PcntCountMode::Decrement => esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE,
+        }
+    }
+}
+
+#[doc = " @brief PCNT channel action on control level"]
+#[derive(Debug, Copy, Clone)]
+pub enum PcntControlMode {
+    #[doc = "< Keep current count mode"]
+    Keep,
+    #[doc = "< Invert current count mode (increase -> decrease, decrease -> increase)"]
+    Reverse,
+    #[doc = "< Hold current count value"]
+    Disable,
+}
+
+impl Into<esp_idf_sys::pcnt_ctrl_mode_t> for PcntControlMode {
+    fn into(self) -> esp_idf_sys::pcnt_ctrl_mode_t {
+        match self {
+            PcntControlMode::Keep => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_KEEP,
+            PcntControlMode::Reverse => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_INVERSE,
+            PcntControlMode::Disable => esp_idf_sys::pcnt_channel_level_action_t_PCNT_CHANNEL_LEVEL_ACTION_HOLD,
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        #[doc = "< PCNT watch point event: threshold1 value event"]
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        #[doc = "< PCNT watch point event: threshold0 value event"]
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        #[doc = "< PCNT watch point event: Minimum counter value"]
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        #[doc = "< PCNT watch point event: Maximum counter value"]
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        #[doc = "< PCNT watch point event: counter value zero event"]
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+pub struct PcntConfig<'d> {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<&'d AnyInputPin>,
+    #[doc = "< Control signal input GPIO number, a negative value will be ignored"]
+    pub ctrl_pin: Option<&'d AnyInputPin>,
+    #[doc = "< PCNT low control mode"]
+    pub lctrl_mode: PcntControlMode,
+    #[doc = "< PCNT high control mode"]
+    pub hctrl_mode: PcntControlMode,
+    #[doc = "< PCNT positive edge count mode"]
+    pub pos_mode: PcntCountMode,
+    #[doc = "< PCNT negative edge count mode"]
+    pub neg_mode: PcntCountMode,
+    #[doc = "< Maximum counter value"]
+    pub counter_h_lim: i16,
+    #[doc = "< Minimum counter value"]
+    pub counter_l_lim: i16,
+    #[doc = "< the PCNT channel"]
+    pub channel: PcntChannel,
+}
+
+#[derive(Debug)]
+pub struct Pcnt {
+    unit: pcnt_unit_t,
+}
+
+impl<'d> Pcnt {
```

Comment:
I'm confused here. Is this a driver, or a peripheral? Since the current release of `esp-idf-hal`, we are strictly separating the notion of a peripheral from the notion of driver. Please check the `CHANGELOG.md` file for more info.) If this is a driver, it needs to take a peripheral (using the `impl Peripheral` magic), and for consistency, it needs to be named `PcntDriver`, where `Pcnt` is the peripheral trait.

Reply:
It's a driver, I think. I implemented the unit allocation the same way that esp-idf v5 works (it has an allocation function and does not have constants for each unit) rather than implementing the peripherial as I initialy was thinking to implemet v4 with a more v5 imterface.  In v5 the channels (two for each unit) are also implemented this way...

Reply:
OK - Implemented as PcntDriver and created Pcnt peripheral.